draft-ietf-mls-protocol.md

```diff
@@ -1243,18 +1243,24 @@ messages refer to KeyPackages for the members being welcomed, and Commits refer
 to Proposals they cover.  These identifiers are computed as follows:
 
 ~~~ tls
-opaque HashReference[16];
+opaque HashReference[Hash.length];
 
 HashReference KeyPackageRef;
 HashReference ProposalRef;
 ~~~
 
 ~~~ pseudocode
-MakeKeyPackageRef(value) = KDF.expand(
-  KDF.extract("", value), "MLS 1.0 KeyPackage Reference", 16)
+MakeKeyPackageRef(value) = RefHash("MLS 1.0 KeyPackage Reference", value)
+MakeProposalRef(value)   = RefHash("MLS 1.0 Proposal Reference", value)
 
-MakeProposalRef(value) = KDF.expand(
-  KDF.extract("", value), "MLS 1.0 Proposal Reference", 16)
+RefHash(label, value) = Hash(RefHashInput)
```

Comment:
Paging @kkohbrok @TWal et al. for crypto review.  @Bren2010 fwiw, I think the more minimal change would be:

``` suggestion
RefHash(label, value) = KDF.expand(KDF.extract("", value), label, KDF.Nh)
```

(That is, just change `16` to `KDF.Nh`.)

From an implementation POV, I don't think there's much difference either way.  Extract/expand is 4 hash invocations instead of 1, but none of this is done at much scale.

Reply:
Using a hash ensures that we're really relying on the ciphersuite's understanding of what a secure hash function is. If KDF.Nh is the size of the output of KDF.Extract (which outputs keys, not collision resistant hashes), then it may again be too small to be collision-secure.

Reply:
At least for the key derivation proofs, we're already assuming that KDFs are collision resistant, which I believe is fairly standard. That being said increasing the length is probably a good idea. 

Also, I don't think that it makes much of a difference if we're using a hash function or a KDF, since we don't rely on any pseudorandomness properties of the hash references.

Reply:
I lean towards using the hash. A hash function is "the" collision resistant primitive and since collision resistance is what we want here, hash would be the right primitive. The main purpose of extract is pseudo-randomness, which is unrelated to collision resistance. (Even if we have to assume collision resistance of extract for some property, using hash here makes it less ambiguous which property is needed.)

Reply:
Thanks @kkohbrok @mulmarta, clearing my objection.